unix, windows: add IPv6 support for uv_udp_multicast_interface

saghul opened the pull request:
/cc @indutny, @bnoordhuis

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit saghul/libuv@c2586eef143f777e9f8e8b91a03cd7bba0ee99b7 has the following error(s):
- First line of commit message must be no longer than 50 characters

The following commiters were not found in the CLA:
- Saúl Ibarra Corretgé

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

saghul commented:
Ok, Jenkins, your turn.

saghul commented:
Sigh, not sure how reliable Jenkins is for this :-/ I manually ran the tests on Linux and Windows and it passes. I'm getting a failure on OSX 10.6, I'll investigate that one...

indutny commented on src/unix/udp.c:
```diff
@@ -610,28 +610,33 @@ int uv_udp_set_multicast_loop(uv_udp_t* handle, int on) {
 int uv_udp_set_multicast_interface(uv_udp_t* handle, const char* interface_addr) {
-  struct in_addr addr;
-  int err;
+  struct sockaddr_in addr4;
```
`struct sockaddr_storage addr_st;` and two pointers to it, instead of two structs?

indutny commented on src/win/udp.c:
```diff
@@ -660,14 +660,12 @@ int uv_udp_set_membership(uv_udp_t* handle,
 int uv_udp_set_multicast_interface(uv_udp_t* handle, const char* interface_addr) {
-  struct in_addr addr;
   int err;
+  struct sockaddr_in addr4;
```
Ditto.

indutny commented:
Few nits, otherwise LGTM

indutny commented:
But recheck jenkins output after fixes ;)

saghul commented:
Will fix, thanks for the review Fedor!

saghul commented:
Fixed nits, I'll test in on OSX and see what happens...

saghul commented:
Ok, I know what the issue is on OSX. When we try to set the interface we are using '::', which will result in having a scope id of 0. Something like 'fe80::1%lo0' would end up having s scope id of 1 (on OSX, because whatever comes after the % symbol is the interface name). On Linux setting the multicast interface to the 0 index (sin6_scope_id) works, but on OSX doesn't, it always gives back EADDRNOTAVAIL, which is surprising, because if you do `getaddrinfo` and try the results in turns, '::1' has a scope id of 0. Fun times.

I'll add an if-def specific to OSX.

Note to self: I'm learning a lot more about IPv6 than I originally intended to.

saghul commented:
@indutny I changed the approach a bit, and fixed the case when you pass NULL, since that should take the default. Can you please review again? Thanks!

saghul commented:
So, even if the man pages say that in order to reset the interface a 0 should be passed: https://developer.apple.com/library/mac/documentation/Darwin/Reference/Manpages/man4/ip6.4.html

OSX decides to return EADDRNOTAVAIL. Seriously, WAT.

bnoordhuis commented:
EADDRNOTAVAIL means that the interface doesn't exist or that it isn't multicast-capable/enabled.  I suspect in your case it's the latter.

saghul commented:
I tried the following: enable multicast on interface with index 1, then try
to reset it by passing a 0 (as per the docs) and I get the error. I should
also mention that if I check the initial state with getsockopt I get a 0
back. So IMHO the behavior OSX is showing here is not kosher.

Or did I miss anything?

Thanks Ben!
On Mar 20, 2014 6:19 PM, "Ben Noordhuis" notifications@github.com wrote:

> EADDRNOTAVAIL can mean that the interface doesn't exist or that it isn't
> multicast-capable/enabled. I suspect in your case it's the latter.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/1206#issuecomment-38195230
> .

bnoordhuis commented:
No, that sounds about right.  Wouldn't be the first xnu kernel bug we have to paper over in libuv.

saghul commented:
Rebased. Can you guys please take a look?

saghul commented:
> No, that sounds about right. Wouldn't be the first xnu kernel bug we have to paper over in libuv.

Do you know where/if that can be reported?

indutny commented:
LGTM

bnoordhuis commented:
> Do you know where/if that can be reported?

https://bugreport.apple.com/‎ - but you need an Apple developer account and it's unlikely that you get actual feedback.

bnoordhuis commented:
Alternatively, you could try the darwin-kernel mailing list - I know at least a handful of Apple kernel engineers read that list.

saghul commented:
Thanks Ben, I'll give those a shot.

saghul commented:
I ended up reporting it :-) http://lists.apple.com/archives/darwin-kernel/2014/Mar/msg00012.html

saghul committed:
unix, windows: add IPv6 support for uv_udp_multicast_interface

saghul commented:
Fixed the test and landed it.
